proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2290 @@
+  ``case_p`` is allowed for a variable ``p``, we can allow arbitrary
+  variables and disallow, in particular, ``case_p``.
+
+  In this case, we would have:
+
+  ::
+
+     map :: (a :(p:+'One)-> b) -> [a] :(p:+'One)-> [b]
+     map f [] = []
+     map f (a:l) = f a : (map f l)
+
+  In practice, under this situation, the type of ``map`` is probably better
+  written as
+
+  ::
+
+     map :: forall p a b q. (p ~ q :+ 'One) => (a :'One-> b) -> [a] :p-> [b]
+
+  In order to play more nicely, for instance, with explicit type
+  applications.
+
+  A benefit is that higher-order functions with no ``case`` such as
+  ``(.)`` are now capable of taking functions with multiplicity ``0`` as
+  argument.
+- A variation on the same idea is to introduce a constraint
+
+  ::
+
+    CaseCompatible :: Multiplicity -> Constraint
+
+  which is discharged automatically by the compiler. Variables
+  implementing this are acceptable in ``case``. So ``map`` would be of
+  type.
+
+  ::
+
+    map :: (CaseCompatible p) => (a :p-> b) -> [a] :p-> [b]
+
+  This is harder to implement than just reusing ``p~q:+'One`` as a
+  constraint, but is more resistant to having more multiplicities than
+  just 0, 1, and ω, as is currently proposed.
+- Another option is to have a type of multiplicities *excluding* ``0``
+  and have another type of extended mulitplicities for multiplicities
+  with ``0``. Note that a different ``(:+)`` and ``(:*)`` would have to
+  act on extended multiplicities.
+
+No annotation on case
+---------------------
+
+.. _`No annotation on case`
+
+Instead of having ``case_p`` (see Multiplicities_) we could just have the
+regular ``case`` (which would correspond to ``case_1`` in this
+proposal's formalism). This would simplify the addition of ``0``.
+
+On the other hand, doing this loses the principle that linear data
+types and unrestricted data types are one and the same. And sacrifices
+much code reuse.
+
+Uniqueness instead of linearity
+-------------------------------
+
+Languages like Clean and Rust have a variant of linear types called
+uniqueness, or ownership, typing. This is a dual notion: instead of
+functions guaranteeing that they use their argument exactly once, and
+no restriction being imposed on the caller, with uniqueness type, the
+caller must guarantee that it has a non-aliased reference to a value,
+and the function has no restriction.
+
+Where uniqueness really shines, is for in-place mutation: the ``write``
+function can take a regular ``Array`` as an argument, it just needs to
+require that it is unique. Freezing is really easy: just drop the
+constraint that the ``Array`` is unique, it will never be writable
+again.
+
+With linear types, we need to have two types ``MArray`` (guaranteed
+unique) and ``Array``, just like in Haskell today. This is fine when
+we are freezing one array: just call ``freeze``. But what if we are
+freezing a list of arrays? Do we need to ``map freeze``? Thi
```

Comment:
```
> :t ($)
($) :: (a -> b) -> a -> b
> :i ($)
($) ::
  forall (r :: GHC.Types.RuntimeRep) a (b :: TYPE r).
  (a -> b) -> a -> b
  	-- Defined in ‘GHC.Base’
infixr 0 $
```

This is about (37) from @goldfirere .

I still don't know what the question is, I'm afraid. Are we indeed talking about the syntax of `case` in the linear core document?

If so the syntax is `case <scrutinee> of z :_p A of { <alts }`. There, `z` is the case binder (for those not familiar with Core, it binds the evaluated form of the scrutiny, it has many uses in Core). The syntax `z :_p A` materialises the multiplicity of `z` (which, in the actual Core implementation are store inside `z` itself). The main reason to materialise it is that the `case_p` from the paper is represented, in Core, as just `case`, but the `p` is taken to be the `p` of the case binder.

Does that clear the misunderstanding? If not, you will have to be even more specific, I'm terribly sorry about this.